src/jitlayers.cpp

```diff
@@ -0,0 +1,396 @@
+    if (jl_options.opt_level>=1)
+        PM->add(createInstructionCombiningPass());   // Clean up after SLP loop vectorizer
+#endif
+#if defined(LLVM35)
+    PM->add(createLoopVectorizePass());         // Vectorize loops
+    PM->add(createInstructionCombiningPass());  // Clean up after loop vectorizer
+#endif
+    //FPM->add(createCFGSimplificationPass());     // Merge & remove BBs
+}
+
+#ifdef USE_ORCJIT
+
+// ------------------------ TEMPORARILY COPIED FROM LLVM -----------------
+// This must be kept in sync with gdb/gdb/jit.h .
+extern "C" {
+
+  typedef enum {
+    JIT_NOACTION = 0,
+    JIT_REGISTER_FN,
+    JIT_UNREGISTER_FN
+  } jit_actions_t;
+
+  struct jit_code_entry {
+    struct jit_code_entry *next_entry;
+    struct jit_code_entry *prev_entry;
+    const char *symfile_addr;
+    uint64_t symfile_size;
+  };
+
+  struct jit_descriptor {
+    uint32_t version;
+    // This should be jit_actions_t, but we want to be specific about the
+    // bit-width.
+    uint32_t action_flag;
+    struct jit_code_entry *relevant_entry;
+    struct jit_code_entry *first_entry;
+  };
+
+  // We put information about the JITed function in this global, which the
+  // debugger reads.  Make sure to specify the version statically, because the
+  // debugger checks the version before we can set it during runtime.
+  extern struct jit_descriptor __jit_debug_descriptor;
+
+  LLVM_ATTRIBUTE_NOINLINE extern void __jit_debug_register_code();
+}
+
+extern JL_DLLEXPORT void ORCNotifyObjectEmitted(JITEventListener *Listener,
+                                      const object::ObjectFile &obj,
+                                      const object::ObjectFile &debugObj,
+                                      const RuntimeDyld::LoadedObjectInfo &L);
+
+namespace {
+
+using namespace llvm;
+using namespace llvm::object;
+using namespace llvm::orc;
+
+/// Do the registration.
+void NotifyDebugger(jit_code_entry* JITCodeEntry) {
+  __jit_debug_descriptor.action_flag = JIT_REGISTER_FN;
+
+  // Insert this entry at the head of the list.
+  JITCodeEntry->prev_entry = nullptr;
+  jit_code_entry* NextEntry = __jit_debug_descriptor.first_entry;
+  JITCodeEntry->next_entry = NextEntry;
+  if (NextEntry) {
+    NextEntry->prev_entry = JITCodeEntry;
+  }
+  __jit_debug_descriptor.first_entry = JITCodeEntry;
+  __jit_debug_descriptor.relevant_entry = JITCodeEntry;
+  __jit_debug_register_code();
+}
+
+// --------------------------------------------------------------------------
+
+class DebugObjectRegistrar {
+private:
+    void NotifyGDB(OwningBinary<ObjectFile> &DebugObj) {
+      const char *Buffer = DebugObj.getBinary()->getMemoryBufferRef().getBufferStart();
+      size_t      Size = DebugObj.getBinary()->getMemoryBufferRef().getBufferSize();
+
+      assert(Buffer && "Attempt to register a null object with a debugger.");
+      jit_code_entry* JITCodeEntry = new jit_code_entry();
+
+      if (!JITCodeEntry) {
+        llvm::report_fatal_error(
```

Comment:
Actually, wouldn't printing a warning most likely allocate, defeating the purpose of this?